Fix: Unignoring directories in .eslintignore (fixes #11684)

belochub opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->
https://github.com/eslint/eslint/issues/11684
<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**

As far as I understand the issue, the filtering of files/directories during globbing must pass absolute paths to the filtering function with the trailing slash present for directories to distinguish them from files, and the relative path extracted from the absolute path in this function must preserve the trailing slash as well since the relative path is passed to `ignore` package afterward, and it can not "unignore" directories that are "marked" as files (have no trailing slash).

~~First of all, I've noticed that the `_mark()`ing in the overridden `GlobSync.prototype._readdir()` method didn't add the trailing slash to the path, IIUC due to the fact that it only uses cache to check if the path points to a directory and this cache is not created until the first call to the parent's `_readdir()` method, thus I've changed the order in which these methods are called.~~

~~Secondly, I've changed the `getRelativePath()` function's behavior to preserve the trailing slash.~~

**Is there anything you'd like reviewers to focus on?**

~~The change to the `getRelativePath()` function is also related to the `IgnoredPaths.prototype.getIgnoredFoldersGlobChecker()` method's behavior and changes it implicitly.~~

~~Also, even though paths are cached inside the `glob.GlobSync` instance, I am not sure if this does not significantly degrade the performance of globbing, I would like to get some help with the benchmarking of this change.~~

EDIT: Some parts of the description above are no longer relevant after https://github.com/eslint/eslint/pull/11546 has landed.

mysticatea commented:
Hi. Thank you for this PR.

I have to tell you an unfortunate thing and a fortunate thing.

Unfortunately, we are rewriting massively about file iteration logic in #11546 and it conflicts with this PR seriously. The new logic doesn't use `GlobSync` class, `glob-utils.js`, and `path-utils.js`.

But fortunately, I have confirmed that #11546 fixes this problem. I will add a test into #11546.

belochub commented:
@mysticatea, I've rebased on `master` and updated the PR to account for the new file iteration logic introduced in https://github.com/eslint/eslint/pull/11546. The original idea behind the PR stayed the same though.

mysticatea commented on lib/cli-engine/file-enumerator.js:
```diff
@@ -296,7 +296,7 @@ class FileEnumerator {
         if (stat && stat.isDirectory()) {
             return this._iterateFilesWithDirectory(
-                absolutePath,
+                absolutePath + path.sep,
```
Would you move `+ path.sep` to L383?

```js
    if (this._isIgnoredFile(directoryPath + path.sep, options)) {
```

Because both cases of a glob and a directory go through that line. We can reduce the fixed lines to the one.

mysticatea commented on lib/util/ignored-paths.js:
```diff
@@ -328,12 +328,16 @@ class IgnoredPaths {
      * @returns {boolean} true if the file path matches one or more patterns, false otherwise
      */
     contains(filepath, category) {
+        const isDir = filepath.slice(-1) === path.sep;
```
I like `.endsWith()` 😃 

```suggestion
        const isDir = filepath.endsWith(path.sep);
```

mysticatea replied:
Ah, also please check both `/` and `\`. [`CLIEngine#isPathIgnored()`](https://github.com/eslint/eslint/blob/aae6f6525557ba06e73f051511646056313fcf91/lib/cli-engine.js#L900-L904) public API can accept both style paths.

belochub replied:
@mysticatea I am not sure I understand the last comment correctly, does that mean that I should modify this condition to mark it as a directory if the path ends with either `/` or `\`?
Wouldn't that break on filenames that end with `\` on OSes that use `/` as a path separator?

mysticatea replied:
My intention was:

```suggestion
        const isDir = filepath.endsWith("/") || filepath.endsWith("\\");
```

Adding trailing slash is fine with `path.sep` because the path is normalized by `path.resolve()`.

belochub replied:
@mysticatea, then I have understood you correctly, but I am still concerned about the following situation:
A user of a system with `/` as a path separator has a file named `./some-name\` (where `\` is a part of the filename) and had chosen to ignore directories named that way, so they put `some-name\/` to their ignore file. If we check for both `/` and `\` at the end to determine if something is a directory, both file `some-name\` and directory `some-name\/` will be passed to the `contains()` method as `some-name\/`, which will lead to ignoring both a file and a directory even though only the directory name was specified in the ignore file.
I understand that this is an extremely unlikely situation, but I just want to make sure that this behavior is okay.

mysticatea replied:
Actually, I don't have knowledge about *NIX file system. But, at least, Windows allows both `/` and `\` as the path separator and we, Windows users, have often used `/`.
